File: Example/Example/ProductViewController.m
//
//  ProductViewController.m
//  Example
//
//  Created by Wyatt Davis on 1/17/23.
//

#import "ProductViewController.h"
#import "ATTNSDKFramework/ATTNSDKFramework-Swift.h"

@interface ProductViewController ()

@end

@implementation ProductViewController
- (void)viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];

  ATTNItem* item = [self buildItem];
  ATTNProductViewEvent* productView = [[ATTNProductViewEvent alloc] initWithItems:@[ item ]];
  productView.deeplink = @"https://mydeeplink.com/products/32432423";

  [[ATTNEventTracker sharedInstance] recordEvent:productView];

  [self showToast:@"Product View event sent" duration:2];
}


- (IBAction)addToCartButtonPressed:(id)sender {
  ATTNItem* item = [self buildItem];
  ATTNAddToCartEvent* addToCart = [[ATTNAddToCartEvent alloc] initWithItems:@[ item ]];

  [[ATTNEventTracker sharedInstance] recordEvent:addToCart];
  [self showToast:@"Add To Cart event sent" duration:2];
}

- (IBAction)addToCartWithDeeplinkButtonPressed:(id)sender {
  ATTNItem* item = [self buildItem];
  ATTNAddToCartEvent* addToCart = [[ATTNAddToCartEvent alloc] initWithItems:@[ item ]];
  addToCart.deeplink = @"https://mydeeplink.com/products/32432423";

  [[ATTNEventTracker sharedInstance] recordEvent:addToCart];
  [self showToast: [NSString stringWithFormat:@"Add To Cart event sent with requestURL(pd): '%@'", addToCart.deeplink] duration:4];
}

- (IBAction)purchaseButtonPressed:(id)sender {
  NSLog(@"Purchase button pressed");

  // Create the Items that were purchased
  ATTNItem* item = [self buildItem];
  // Create the Order
  ATTNOrder* order = [[ATTNOrder alloc] initWithOrderId:@"778899"];
  // Create PurchaseEvent
  ATTNPurchaseEvent* purchase = [[ATTNPurchaseEvent alloc] initWithItems:@[ item ] order:order];

  [[ATTNEventTracker sharedInstance] recordEvent:purchase];

  [self showToast:@"Purchase event sent" duration:2];
}

- (ATTNItem*)buildItem {
  // Build Item with required fields
  ATTNItem* item = [[ATTNItem alloc] initWithProductId:@"222" productVariantId:@"55555" price:[[ATTNPrice alloc] initWithPrice:[[NSDecimalNumber alloc] initWithString:@"15.99"] currency:@"USD"]];
  // Add some optional fields
  item.name = @"T-Shirt";
  item.category = @"Tops";
  return item;
}

- (IBAction)customEventButtonPressed:(id)sender {
  ATTNCustomEvent* customEvent = [[ATTNCustomEvent alloc] initWithType:@"Added to Wishlist" properties:@{@"wishlistName" : @"Gift Ideas"}];

  [[ATTNEventTracker sharedInstance] recordEvent:customEvent];

  [self showToast:@"Custom event sent" duration:2];
}

- (void)showToast:(NSString*)message duration:(int)duration {
  UIAlertController* alert = [UIAlertController alertControllerWithTitle:nil
                                                                 message:message
                                                          preferredStyle:UIAlertControllerStyleAlert];

  [self presentViewController:alert animated:YES completion:nil];

  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, duration * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
    [alert dismissViewControllerAnimated:YES completion:nil];
  });
}

@end

File: ExampleSwift/ExampleSwift/ProductViewController.swift
//
//  ProductViewController.swift
//  ExampleSwift
//
//  Created by Wyatt Davis on 2/9/23.
//

import Foundation
import os
import UIKit
import ATTNSDKFramework

class ProductViewController : UIViewController {
    @IBOutlet var addToCartBtn : UIButton?
    @IBOutlet var purchaseBtn : UIButton?
    @IBOutlet weak var customEventBtn: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let item : ATTNItem = buildItem()
        let productView : ATTNProductViewEvent = ATTNProductViewEvent(items: [item])
        ATTNEventTracker.sharedInstance().record(productView)
        
        showToast(message: "Product View event sent")
    }
    
    @IBAction func addToCartBtnPressed(sender: Any) {
        let item : ATTNItem = self.buildItem()
        let addToCart : ATTNAddToCartEvent = ATTNAddToCartEvent(items: [item])
        ATTNEventTracker.sharedInstance().record(addToCart)
        
        showToast(message: "Add To Cart event sent")
    }
    
    @IBAction func purchaseBtnPressed(sender: Any) {
        // Create the Items that were purchased
        let item : ATTNItem = self.buildItem()
        // Create the Order
        let order : ATTNOrder = ATTNOrder(orderId: "778899")
        // Create PurchaseEvent
        let purchase : ATTNPurchaseEvent = ATTNPurchaseEvent(items: [item], order: order)
        // Send the PurchaseEvent
        ATTNEventTracker.sharedInstance().record(purchase)
        
        showToast(message: "Purchase event sent")
    }
    
    @IBAction func customEventBtnPressed(sender: Any) {
        let customEvent : ATTNCustomEvent = ATTNCustomEvent(type: "Added to Wishlist", properties: ["wishlistName": "Gift Ideas"]);
        ATTNEventTracker.sharedInstance().record(customEvent)
        
        showToast(message: "Custom event sent")
    }

    func buildItem() -> ATTNItem {
        // Build Item with required fields
        let item : ATTNItem = ATTNItem(productId: "222", productVariantId: "55555", price: ATTNPrice(price: NSDecimalNumber(string: "15.99"), currency: "USD"))
        
        // Add some optional fields
        item.name = "T-Shirt"
        item.category = "Tops"
        
        return item
    }
    
    func showToast(message: String) {
        let alert : UIAlertController = UIAlertController(title: nil, message: message, preferredStyle: .alert)
        
        self.present(alert, animated: true)
        
        let deadline = DispatchTime.now() + .seconds(1)
        
        DispatchQueue.main.asyncAfter(deadline: deadline) {
            alert.dismiss(animated: true)
        }
    }
}
